text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+data into chunks, each of which would be independently mutable; we could either
+do this explicitly (allocating an array of disjoint slices) or via some other
+mechanism, similar to a `RefCell`, where we would keep track at runtime of which regions
+were currently borrowed.
+
+This would work (although it would add significant complexity) in cases where we
+had foreknowledge of the position and size of an object, but I do not believe
+this could work when dealing with tables that host offsets, or at least not in
+the general case. The problem with offsets is that they are not verified ahead
+of time, and they often point to objects that do not have a known size. This
+means there is really no way to know exactly how much memory a table at a given
+offset is going to access. While it is possible to imagine complicated runtime
+schemes for tracking this, it would definitely require allocating additional
+data structures, and would significantly degrade the ergonomics of the API, and
+significantly increase the learning curve.
+
+I've been thinking about this problem quite a bit, and my opinion it is evidence
+of a fundamental tension in the design goals of this project.
+
+#### deciding on goals
+
+I believe that we currently have two conflicting goals. One of the rationale for
+this project is concern about the difficulty of getting developers up to speed
+on HarfBuzz, and how hard it is to add features there without accidentally
+introducing bugs. There is a hope that Rust's safety guarantees can give us a
+code base that is easier to modify and hack on; however there is *also* a desire
+to end up with code that is more-or-less equivalent to the existing C++.
+
+The basic tension is this: the ergonomics and safety advantages that Rust offers
+work by explicitly preventing certain behaviour, such as the aliasing of mutable
+variables. We cannot translate idiomatic C++ directly into Rust and get safety;
+we get safety by ruling out certain patterns.
+
+I think it is really worth stepping back and considering our goals. We
+would like to have a tool for authoring & manipulating font files, and we would
+like it to be a) fast and b) easy to use. This is something that we can achieve,
+but I do not think the result is going to look like it might look in C++. My
+feeling is that too much of our thinking (or at least my thinking!) so far has
+been focused on porting an existing design to Rust, and that this may be
+misguided.
+
+#### Subsetting is special
+
+A major focus of (and motivation for) this work has been hb-subset. I think this
+has influenced our thinking in important ways. I'm not entirely confident in my
+analysis here, so please correct any bad assumptions. With that said:
+
+- My sense is that the design of hb-subset is largely path-dependent on the
+  existing design of HarfBuzz. This is just an intuition. What this means to me,
+  though, is that if we were designing a subsetter from scratch, we might make
+  different choices.
```

Comment:
To my understanding hb-subset is based on what we learned from FontTools subsetting. I'm not aware of any major choices in hb-subset I'd expect to make differently if starting clean.

@garretrieger wdyt?

Reply:
On this point I definitely don't have any strong intuitions or beliefs, it's more something I'm curious about.